documentation/API-Testing-Guidelines.md

```diff
@@ -1,41 +1,288 @@
+
+Test plan must validate that providing a valid input, a success response is received and that the response body complies with the JSON-schema in the API spec.
+
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
+
+For path resources with CRUD (Create/Read/Update/Delete) operations, some scenario should validate that the new state of the resource is coherent with the response to the operation.
+
+* For Create operations:
+  - If operation success returns the created object, validate that the properties of the response object are coherent with the request object
+  - Validate that the created object can be read by the GET operation for the item id.
 
-## Best practices and recommendations <a name="recommendations"></a>
+* For Update operations:
+  - If operation success returns the updated object, validate that the properties of the response object are coherent with the request object
+  - Validate that the updated object can be read by the GET operation for the item id, with the modified properties.
+
+* For Delete operations:
+  - Validate that trying to read the successfully deleted object, with the GET operation for the deleted item id, returns the error specified in the API spec, typically 404.
+
+### Notifications <a name="notifications"></a>
+
+For operation with implicit subscriptions:
+
+* Check that when a webhook is provided, the expected events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
+* If the API allows to update a previously provided webhook:
+  - If the `notificationUrl` is modified, validate that events are received in the modified value.
+  - If the `notificationUrl` can be nullified, validate that events are not longer received.
+  - If the `notificationAuthToken` is modified, validate that the new token is used for the events.
+
+For the explicit subscriptions model:
+
+* Validate that the subscribed events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
+* For subscriptions that provide `subscriptionExpireTime`, validate that the subscribed events are not longer received after the expiration time.
+* For subscriptions that provide `subscriptionMaxEvents`, validate that the subscribed events are not longer received after the maximum events limit is reached..
+* Validate that after a subscription is deleted, the subscribed events are not longer received.
+
+### Error scenarios <a name="error-scenarios"></a>
+
+* All errors explicitly documented in the API spec must be covered by one or more dedicated error scenarios.
+  - In particular operations that expect a `device` object allowing one or more identifiers to be provided, must include scenarios to test all error cases defined. 
```

Comment:
@jlurien Can you please clarify why only device object is considered explicitly? We think generic phrase would be fine for covering error scenarios.

Reply:
Because recently errors for device processing have been discussed specifically, and new design guidelines also have a dedicated mention to device error management. But this do not exclude any other error scenario. Idea is to give an example of deep of tests that are expected,

Reply:
added reference to the device error management section of the API design guidelines